knative-operator/pkg/common/telemetry/telemetry.go

```diff
@@ -0,0 +1,186 @@
+package telemetry
+
+import (
+	"context"
+	"sync"
+
+	"github.com/openshift-knative/serverless-operator/knative-operator/pkg/common"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	kafkasourcev1beta1 "knative.dev/eventing-contrib/kafka/source/pkg/apis/sources/v1beta1"
+	eventingsourcesv1beta1 "knative.dev/eventing/pkg/apis/sources/v1beta1"
+	servingv1 "knative.dev/serving/pkg/apis/serving/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+var log = common.Log.WithName("telemetry")
+
+type Telemetry struct {
+	name            string
+	stop            chan struct{}
+	metricsSnapshot sync.Map
+	tc              *controller.Controller
+}
+
+func NewTelemetry(name string, mgr manager.Manager, objects []runtime.Object) (*Telemetry, error) {
+	t := &Telemetry{}
+	t.name = name
+	t.stop = make(chan struct{})
+	tc, err := newTelemetryController(name, objects, mgr, t)
+	if err != nil {
+		return nil, err
+	}
+	t.tc = tc
+	return t, nil
+}
+
+// TryStartTelemetry setups telemetry per component either Eventing, KnativeKafka or Serving.
+// When called it assumes that the component has status ready.
+func (t *Telemetry) TryStartTelemetry(api client.Client, mgr manager.Manager) error {
+	log.Info("starting telemetry for:", "component", t.name)
+	// Initialize metrics before we start the corresponding controller.
+	// There is a tiny window to miss events here, but should be ok for telemetry purposes.
+	t.InitializeAndTakeMetricsSnapshot(api)
+	// Start our controller in a goroutine so that we do not block.
+	if t.tc != nil { // meant to allow nop objects in tests
```

Comment:
Shouldn't this be a `if t == nil { return }` at the beginning of the function?

Reply:
It could be but it should be a bug if this happens. At least that is the common practice about whether to check pointers in receiver methods.

Reply:
I was actually saying that because we rely on the zero value (i.e. nil) in tests apparently. We use that pattern in other methods on this type already.

Reply:
I will change in all functions anyway. I could not parse it above because if I do it for one I should do it for all imho.

Reply:
Yeah I agree, so i didn't parse why we did it different here than in the others.

Reply:
Because I read that you should not check the pointer in receiver methods as a practice. Anyway will do it.

Reply:
But you do it already?